cpp/src/arrow/compare.cc

```diff
@@ -1550,7 +1551,31 @@ bool ArrayStatisticsValueTypeEquals(
         return v1 == v2;
       }
       // It is unreachable
-      DCHECK(false);
+      Unreachable("The types are different.");
+      return false;
+    };
+    return std::visit(EqualsVisitor, left.value(), right.value());
+  }
+}
+
+bool ArrayStatisticsNumericTypeEquals(std::optional<ArrayStatistics::NumericType> left,
+                                      std::optional<ArrayStatistics::NumericType> right,
+                                      const EqualOptions& options) {
```

Comment:
It seems that there is no function body difference between this and `ArrayStatisticsValueTypeEquals()`.

Can we unify them as a template function?